### ISSUE TITLE
Случайное дублирование заказов

### ISSUE THREAD

**Issue by AlexSus** [modx-pro/miniShop2#425]
### Резюме

Уже не первый раз замечаю, что при невыясненных пока обстоятельствах покупатель может оформить заказ несколько раз. Пытался как-то отследить, найти закономерности, но ничего не получилось. При этом у дублированных заказов может оказаться один и тот же номер заказа, а иногда последовательно создаются несколько абсолютно одинаковых заказов, но с разными номерами. Тоже не нашел закономерности. Может у кого-то были похожие ситуации? Очень хочется понять где кроется проблема.
Вот буквально вчера снова была такая проблема:

![изображение](https://user-images.githubusercontent.com/24259648/82144955-d629a000-9871-11ea-8465-db89deb56bc1.png)

### Шаг для воспроизведения

Воспроизвести не получилось, пробовал сам делать заказы, из разных браузеров, с телефона, с разными условиями, искусственно повышал пинг, ограничивал скорость интернета.
Грешил на ошибки в скриптах, убрал строчку, которая в минишопе при отправке заказа добавляет кнопке состояние disabled. Думал может как-то люди умудряются несколько раз щелкнуть несмотря на это, но нет, в таком случае все следующие заказы получаются пустыми, а не абсолютными копиями. 

### Среда

miniShop2 2.4.18-pl, версия MODX Revolution 2.7.1-pl,
стоит модуль mspYaCassa 2.0.12-beta

**Comment by AlexSus** [comment]
Может у кого есть идеи как это можно попробовать отдебажить, буду очень признателен любым идеям.
Причём грешил ещё на хостинг, но месяц назад сменил хостинг, а проблема осталась.

**Comment by Ruslan-Aleev** [comment]
Может какой-то компонент работает еще с заказами miniShop2 и он такое делает? Сколько работал с чистым miniShop2 - не замечал дублирования заказов.

**Comment by CrazyBoy49z** [comment]
Тоже не замечал такой проблеми, попробуйте может при отправке возникает какая ошибка, или проблема в стороннему плагине, или редирект не срабатывает, и пользователь нажимает F5

**Comment by Electrica** [comment]
Можно воспроизвести, если кнопку нажать дважды. Когда какая то ошибка не дает кнопки стать disabled

**Comment by AlexSus** [comment]
> 
> 
> Можно воспроизвести, если кнопку нажать дважды. Когда какая то ошибка не дает кнопки стать disabled

Нет, не получается так воспроизвести. Когда я пытался повторить проблему, я специально убрал эту строчку, отвечающую за то, чтобы кнопку нельзя было нажать несколько раз:

![изображение](https://user-images.githubusercontent.com/24259648/82150927-024e1c80-9884-11ea-87f7-f62cdd4d2921.png)

В итоге создавался первый реальный заказ, как должно, а затем 2-4 заказов с нулевой стоимостью и без товаров. Что в принципе и логично, т.к. после создания заказа происходит сброс сессии, и уже нет товаров в корзине, значит все работает правильно.

**Comment by AlexSus** [comment]
> 
> 
> Может какой-то компонент работает еще с заказами miniShop2 и он такое делает? Сколько работал с чистым miniShop2 - не замечал дублирования заказов.

Тоже подумал, что может компонент какой-то влезает, но вспомнил, что есть другой магазин, где минимум сторонних компонентов и нет ни одного, который вмешивается ни в рассчет доставок, ни в оплату. Просто используются стандартные доставки минишопа и стандартные оплаты. Там я думал если не найду таких дублирований, то значит проблема в компоненте каком-то, но нет, нашел и там такие заказы. Здесь кстати еще страннее, тут у второго заказа нет прикрепленного пользователя, а у третьего пустая корзина.

![изображение](https://user-images.githubusercontent.com/24259648/82670817-099c6e00-9c68-11ea-8705-4ddcdaac949f.png)

Если есть у вас более-менее активные проекты на минишопе, уделите, пожалуйста, 5-10 минут, пробежитесь по заказам. Хочется понять, есть ли реально проблема из коробки или нет.

**Comment by rebroff** [comment]
Та же проблема. С периодичностью раз в два-три месяца. Заказ дублируется иногда до 8 раз. Причем 2-3 из повторов с пустой корзиной. Время заказа всех дублей совпадает с точностью до минуты.

Тоже пытался воспроизвести. Перепробовал все комбинации доставок и оплат, пытался смоделировать любое возможное поведение пользователя типа многократных кликов, обновления страницы, отключения Интернета - без толку.
